protocols/noise/src/io.rs

```diff
@@ -255,29 +235,25 @@ impl<T: AsyncRead + Unpin> AsyncRead for NoiseOutput<T> {
 }
 
 impl<T: AsyncWrite + Unpin> AsyncWrite for NoiseOutput<T> {
-    fn poll_write(
-        mut self: Pin<&mut Self>,
-        cx: &mut Context<'_>,
-        buf: &[u8],
-    ) -> Poll<Result<usize, std::io::Error>>{
+    fn poll_write(mut self: Pin<&mut Self>, cx: &mut Context, buf: &[u8]) -> Poll<io::Result<usize>> {
         let mut this = self.deref_mut();
-
-        let buffer = this.buffer.borrow_mut();
-
         loop {
             trace!("write state: {:?}", this.write_state);
             match this.write_state {
                 WriteState::Init => {
                     this.write_state = WriteState::BufferData { off: 0 }
                 }
                 WriteState::BufferData { ref mut off } => {
-                    let n = std::cmp::min(MAX_WRITE_BUF_LEN - *off, buf.len());
-                    buffer.write[*off .. *off + n].copy_from_slice(&buf[.. n]);
+                    let n = min(MAX_WRITE_BUF_LEN, this.write_buffer.len().saturating_add(buf.len()));
```

Comment:
Do I understand correctly, that we never set `this.write_buffer` back to `0`. Let's say one always calls `poll_write` with a very small `buf` (e.g. length of 1). In addition let's say one always calls `poll_flush` right afterwards each time.

As far as I can tell `this.write_buffer` would eventually be as large as `MAX_WRITE_BUF_LEN` due to `this.write_buffer.len().saturating_add(buf.len())` followed by `this.write_buffer.resize(n, 0u8);` even though one only ever needed `this.write_buffer` to be of length 1.

Would resizing `this.write_buffer` with the current offset (`off`) instead of `this.write_buffer.len()` solve this?

```suggestion
                    let n = min(MAX_WRITE_BUF_LEN, off.saturating_add(buf.len()));
```

Reply:
Good catch!